In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")
# Dispaly max column
pd.pandas.set_option('display.max_columns',None)
# Display max rows
#pd.pandas.set_option('display.max_rows', None)

In [2]:
## loading dataset using pandas library
telcom = pd.read_csv("telcom_clean.csv")

## print the top5 records
telcom.head(5)

Bearer Id                Start  Start ms                  End  End ms  \
0  1.311450e+19  2019-04-04 12:01:00     770.0  2019-04-25 14:35:00   662.0   
1  1.311450e+19  2019-04-09 13:04:00     235.0  2019-04-25 08:15:00   606.0   
2  1.311450e+19  2019-04-09 17:42:00       1.0  2019-04-25 11:58:00   652.0   
3  1.311450e+19  2019-04-10 00:31:00     486.0  2019-04-25 07:36:00   171.0   
4  1.311450e+19  2019-04-12 20:10:00     565.0  2019-04-25 10:40:00   954.0   

   Dur. (ms)          IMSI  MSISDN/Number          IMEI Last Location Name  \
0  1823652.0  2.082010e+14   3.366496e+10  3.552120e+13        9.16457E+15   
1  1365104.0  2.082020e+14   3.368185e+10  3.579400e+13            L77566A   
2  1361762.0  2.082000e+14   3.376063e+10  3.528150e+13            D42335A   
3  1321509.0  2.082010e+14   3.375034e+10  3.535660e+13            T21824A   
4  1089009.0  2.082010e+14   3.369980e+10  3.540700e+13            D88865A   

   Avg RTT DL (ms)  Avg RTT UL (ms)  Avg Bearer TP DL (kbps)  \
0        42.000000         5.000000                     23.0   
1        65.000000         5.000000                     16.0   
2       109.795706        17.662883                      6.0   
3       109.795706        17.662883                     44.0   
4       109.795706        17.662883                      6.0   

   Avg Bearer TP UL (kbps)  TCP DL Retrans. Vol (Bytes)  \
0                     44.0                 2.080991e+07   
1                     26.0                 2.080991e+07   
2                      9.0                 2.080991e+07   
3                     44.0                 2.080991e+07   
4                      9.0                 2.080991e+07   

   TCP UL Retrans. Vol (Bytes)  DL TP < 50 Kbps (%)  \
0                759658.664811                100.0   
1                759658.664811                100.0   
2                759658.664811                100.0   
3                759658.664811                100.0   
4                759658.664811                100.0   

   50 Kbps < DL TP < 250 Kbps (%)  250 Kbps < DL TP < 1 Mbps (%)  \
0                             0.0                            0.0   
1                             0.0                            0.0   
2                             0.0                            0.0   
3                             0.0                            0.0   
4                             0.0                            0.0   

   DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  10 Kbps < UL TP < 50 Kbps (%)  \
0                 0.0                100.0                            0.0   
1                 0.0                100.0                            0.0   
2                 0.0                100.0                            0.0   
3                 0.0                100.0                            0.0   
4                 0.0                100.0                            0.0   

   50 Kbps < UL TP < 300 Kbps (%)  UL TP > 300 Kbps (%)  HTTP DL (Bytes)  \
0                             0.0                   0.0     1.144710e+08   
1                             0.0                   0.0     1.144710e+08   
2                             0.0                   0.0     1.144710e+08   
3                             0.0                   0.0     1.144710e+08   
4                             0.0                   0.0     1.144710e+08   

   HTTP UL (Bytes)  Activity Duration DL (ms)  Activity Duration UL (ms)  \
0     3.242301e+06                    37624.0                    38787.0   
1     3.242301e+06                      168.0                     3560.0   
2     3.242301e+06                        0.0                        0.0   
3     3.242301e+06                     3330.0                    37882.0   
4     3.242301e+06                        0.0                        0.0   

    Dur. (ms).1 Handset Manufacturer                 Handset Type  \
0  1.823653e+09              Samsung   Samsung Galaxy A5 Sm-A520F   
1  1.365104e+09              Samsung  Samsung Galaxy J5 (Sm-J530)  

In [3]:
telcom.isnull().sum()

Bearer Id                                   0
Start                                       0
Start ms                                    0
End                                         0
End ms                                      0
Dur. (ms)                                   0
IMSI                                        0
MSISDN/Number                               0
IMEI                                        0
Last Location Name                          0
Avg RTT DL (ms)                             0
Avg RTT UL (ms)                             0
Avg Bearer TP DL (kbps)                     0
Avg Bearer TP UL (kbps)                     0
TCP DL Retrans. Vol (Bytes)                 0
TCP UL Retrans. Vol (Bytes)                 0
DL TP < 50 Kbps (%)                         0
50 Kbps < DL TP < 250 Kbps (%)              0
250 Kbps < DL TP < 1 Mbps (%)               0
DL TP > 1 Mbps (%)                          0
UL TP < 10 Kbps (%)                         0
10 Kbps < UL TP < 50 Kbps (%)     

In [4]:
telcom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   Bearer Id                                 150000 non-null  float64
 1   Start                                     150000 non-null  object 
 2   Start ms                                  150000 non-null  float64
 3   End                                       150000 non-null  object 
 4   End ms                                    150000 non-null  float64
 5   Dur. (ms)                                 150000 non-null  float64
 6   IMSI                                      150000 non-null  float64
 7   MSISDN/Number                             150000 non-null  float64
 8   IMEI                                      150000 non-null  float64
 9   Last Location Name                        150000 non-null  object 
 10  Avg RTT DL (ms)     

In [5]:
telcom.describe()

Bearer Id       Start ms         End ms     Dur. (ms)          IMSI  \
count  1.500000e+05  150000.000000  150000.000000  1.500000e+05  1.500000e+05   
mean   1.013887e+19     499.188200     498.800880  1.046086e+05  2.082017e+14   
std    2.883606e+18     288.611834     288.097653  8.103762e+04  2.145208e+10   
min    6.917540e+18       0.000000       0.000000  7.142000e+03  2.040470e+14   
25%    7.349880e+18     250.000000     251.000000  5.744050e+04  2.082010e+14   
50%    7.349880e+18     499.000000     500.000000  8.639900e+04  2.082020e+14   
75%    1.304240e+19     749.000000     750.000000  1.324302e+05  2.082020e+14   
max    1.318650e+19     999.000000     999.000000  1.859336e+06  2.140740e+14   

       MSISDN/Number          IMEI  Avg RTT DL (ms)  Avg RTT UL (ms)  \
count   1.500000e+05  1.500000e+05    150000.000000    150000.000000   
mean    4.188282e+10  4.847454e+13       109.795706        17.662883   
std     2.438739e+12  2.237367e+13       559.344484        76.530188   
min     3.360100e+10  4.400150e+11         0.000000         0.000000   
25%     3.365151e+10  3.546260e+13        35.000000         3.000000   
50%     3.366380e+10  3.572240e+13        54.000000         7.000000   
75%     3.368473e+10  8.611970e+13       109.795706        17.662883   
max     8.823970e+14  9.900120e+13     96923.000000      7120.000000   

       Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
count            150000.000000            150000.000000   
mean              13300.045927              1770.428647   
std               23971.878541              4625.355500   
min                   0.000000                 0.000000   
25%                  43.000000                47.000000   
50%                  63.000000                63.000000   
75%               19710.750000              1120.000000   
max              378160.000000             58613.000000   

       TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
count                 1.500000e+05                 1.500000e+05   
mean                  2.080991e+07                 7.596587e+05   
std                   1.172360e+08                 1.577621e+07   
min                   2.000000e+00                 1.000000e+00   
25%                   1.332913e+06                 6.300775e+04   
50%                   2.080991e+07                 7.596587e+05   
75%                   2.080991e+07                 7.596587e+05   
max                   4.294426e+09                 2.908226e+09   

       DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
count        150000.000000                   150000.000000   
mean             92.844754                        3.069355   
std              13.005265                        6.199613   
min               0.000000                        0.000000   
25%              91.000000                        0.000000   
50%             100.000000                        0.000000   
75%             100.000000                        4.000000   
max             100.000000                       93.000000   

       250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
count                  150000.000000       150000.000000        150000.000000   
mean                        1.717341            1.609654            98.530142   
std                         4.149084            4.816754             4.622050   
min                         0.000000            0.000000             0.000000   
25%                         0.000000            0.000000            99.000000   
50%                         0.000000            0.000000           100.000000   
75%                         1.000000            1.000000           100.000000   
max                       100.000000           94.000000           100.000000   

       10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
count                  150000.000000                   150000.000000   
mean                        0.776749                        0.

In [6]:
data = telcom.copy()
data['Total Traffic (Bytes)']= data['Total DL (Bytes)'] + data["Total UL (Bytes)"]
data["Total Youtube Volume(Bytes)"]= data["Youtube DL (Bytes)"] + data["Youtube UL (Bytes)"]
data["Total Google Volume(Bytes)"]= data["Google DL (Bytes)"] + data["Google UL (Bytes)"]
data["Total Email Volume(Bytes)"]= data["Email DL (Bytes)"] + data["Email UL (Bytes)"]
data["Total Social Media Volume(Bytes)"]= data["Social Media DL (Bytes)"] + data["Social Media UL (Bytes)"]
data["Total Netflix Volume(Bytes)"]= data["Netflix DL (Bytes)"] + data["Netflix UL (Bytes)"]
data["Total Gaming Volume(Bytes)"]= data["Gaming DL (Bytes)"] + data["Gaming UL (Bytes)"]
data["Total Other Volume(Bytes)"]= data["Other DL (Bytes)"] + data["Other UL (Bytes)"]

In [7]:
## Aggregate total data volume (in Bytes) during this session for each application
apps_data_volume = data.groupby(['Dur. (ms)','IMSI', 'MSISDN/Number']).agg({
    'Total Traffic (Bytes)': 'sum',
    'Total Youtube Volume(Bytes)': 'sum',
    'Total Google Volume(Bytes)': 'sum',
    'Total Email Volume(Bytes)': 'sum',
    'Total Social Media Volume(Bytes)': 'sum',
    'Total Netflix Volume(Bytes)': 'sum',
    'Total Gaming Volume(Bytes)': 'sum',
    'Total Other Volume(Bytes)': 'sum',
}).reset_index()

apps_data_volume.head()

Dur. (ms)          IMSI  MSISDN/Number  Total Traffic (Bytes)  \
0     7142.0  2.082020e+14   3.376351e+10            558334893.0   
1     7146.0  2.082010e+14   3.366566e+10            678517390.0   
2     7189.0  2.082020e+14   3.365933e+10            279804529.0   
3     7212.0  2.082017e+14   4.188282e+10            140464511.0   
4     7258.0  2.082010e+14   3.369839e+10            836659601.0   

   Total Youtube Volume(Bytes)  Total Google Volume(Bytes)  \
0                   15264002.0                   8933196.0   
1                   19778800.0                   4680503.0   
2                   22378059.0                   5458634.0   
3                   18559695.0                   3322061.0   
4                   35149499.0                  10015999.0   

   Total Email Volume(Bytes)  Total Social Media Volume(Bytes)  \
0                  2264693.0                         1065060.0   
1                  1075475.0                         1163449.0   
2                  1028980.0                         3431790.0   
3                   925210.0                         2810974.0   
4                  1629885.0                         2492057.0   

   Total Netflix Volume(Bytes)  Total Gaming Volume(Bytes)  \
0                    7859901.0                 517329331.0   
1                   31358730.0                 619324450.0   
2                   32348712.0                 203674294.0   
3                   22971917.0                  91126558.0   
4                   13639254.0                 767565928.0   

   Total Other Volume(Bytes)  
0                 33455346.0  
1                375530262.0  
2                783669544.0  
3                137512067.0  
4                228133871.0

### Sessions frequency

In [8]:
# Calculate the sessions frequency for each user
sessions_frequency = telcom['MSISDN/Number'].value_counts()

sessions_frequency = pd.DataFrame({'Sessions Frequency': sessions_frequency}).reset_index()

# Display the head of the sessions frequency series
sessions_frequency.head()

MSISDN/Number  Sessions Frequency
0   4.188282e+10                1065
1   3.362632e+10                  18
2   3.361489e+10                  17
3   3.362578e+10                  17
4   3.365973e+10                  16

***The table above shows the top five users with the highest session frequency in the dataset. The session frequency is the count of sessions per user, identified by their MSISDN/Number. The user with the highest number of sessions has 1065 sessions, followed by one user with 18 sessions, and so on.***

## The Duration of the session

In [9]:
# Calculate the duration of each session
session_durations = telcom.groupby('MSISDN/Number')['Dur. (ms)'].sum()

## Convert the result into a DataFrame
session_durations = pd.DataFrame({'Sessions Duration (ms)': session_durations}).reset_index()

# Display the head of the session durations series
session_durations.head(10)

MSISDN/Number  Sessions Duration (ms)
0   3.360100e+10                116720.0
1   3.360100e+10                181230.0
2   3.360100e+10                134969.0
3   3.360101e+10                 49878.0
4   3.360101e+10                 37104.0
5   3.360101e+10                253983.0
6   3.360101e+10                128360.0
7   3.360101e+10                 86399.0
8   3.360101e+10                495702.0
9   3.360102e+10                124854.0

***The table above shows the total duration of sessions for the top five users in the dataset, measured in milliseconds. Each row represents a unique user identified by their MSISDN/Number and the corresponding sum of durations for all sessions they had.***

## The session total traffic (download and upload (bytes))


In [10]:
# Calculate the total traffic (download and upload in bytes) for each session
session_total_traffic = telcom.groupby('MSISDN/Number')[['Total DL (Bytes)', 'Total UL (Bytes)']].sum()

# Add a new column for the total traffic per session
session_total_traffic['Total Traffic (Bytes)'] = session_total_traffic['Total DL (Bytes)'] + session_total_traffic['Total UL (Bytes)']

# Display the head of the session total traffic dataframe
session_total_traffic

Total DL (Bytes)  Total UL (Bytes)  Total Traffic (Bytes)
MSISDN/Number                                                           
3.360100e+10       8.426375e+08      3.605311e+07           8.786906e+08
3.360100e+10       1.207552e+08      3.610446e+07           1.568596e+08
3.360100e+10       5.566597e+08      3.930682e+07           5.959665e+08
3.360101e+10       4.019932e+08      2.032753e+07           4.223207e+08
3.360101e+10       1.363130e+09      9.428053e+07           1.457411e+09
...                         ...               ...                    ...
3.379000e+10       4.445751e+08      3.573224e+07           4.803073e+08
4.188282e+10       4.869236e+11      4.432530e+10           5.312489e+11
3.197020e+12       1.948281e+08      3.729592e+07           2.321240e+08
3.370000e+14       5.396350e+08      5.665284e+07           5.962878e+08
8.823970e+14       7.869760e+07      6.045605e+07           1.391536e+08

[106857 rows x 3 columns]

***The table displays the total download and upload traffic for the top five users, along with their combined total traffic.***

# Task 2.1 - Based on the above submit the Python script and slide:

## Aggregate the above metrics per customer id (MSISDN) and report the top 10 customers per engagement metric 


In [11]:
engagment_metrics = data.groupby('MSISDN/Number').agg({'Bearer Id': 'count','Dur. (ms)': 'sum', 'Total Traffic (Bytes)': 'sum',})
engagment_metrics = engagment_metrics.rename(columns={'Bearer Id': 'Sessions Frequency', 'Dur. (ms)': 'Sessions Duration (ms)'})
engagment_metrics.sort_values(by=['Sessions Frequency'], ascending=False).head(10)

Sessions Frequency  Sessions Duration (ms)  \
MSISDN/Number                                               
4.188282e+10                 1065              72446396.0   
3.362632e+10                   18               8791927.0   
3.362578e+10                   17              18553754.0   
3.361489e+10                   17               9966898.0   
3.365973e+10                   16               4035428.0   
3.367588e+10                   15               4865947.0   
3.376054e+10                   15               9279434.0   
3.366716e+10                   13               8744914.0   
3.360452e+10                   12               5207990.0   
3.362708e+10                   12               4703516.0   

               Total Traffic (Bytes)  
MSISDN/Number                         
4.188282e+10            5.312489e+11  
3.362632e+10            7.971167e+09  
3.362578e+10            8.499621e+09  
3.361489e+10            8.846226e+09  
3.365973e+10            7.705863e+09  
3.367588e+10            7.891111e+09  
3.376054e+10            8.514774e+09  
3.366716e+10            5.618394e+09  
3.360452e+10            5.487855e+09  
3.362708e+10            5.754731e+09

### The top 10 customers have been identified for each engagement metric:

- Session frequency: The number of sessions each customer had.
- Session duration: The total duration of all sessions per customer.
- Total traffic: The total data (download and upload) consumed per customer.

In [12]:
engagment_metrics.describe()

Sessions Frequency  Sessions Duration (ms)  Total Traffic (Bytes)
count       106857.000000            1.068570e+05           1.068570e+05
mean             1.403745            1.468438e+05           6.959272e+08
std              3.352062            2.892198e+05           1.695706e+09
min              1.000000            7.142000e+03           3.324901e+07
25%              1.000000            7.130900e+04           3.585530e+08
50%              1.000000            1.027410e+05           6.179232e+08
75%              2.000000            1.727990e+05           8.574358e+08
max           1065.000000            7.244640e+07           5.312489e+11

## Normalize each engagement metric and run a k-means (k=3) to classify customers into three groups of engagement.

In [20]:
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans

# k means clustering
kmeans = KMeans(n_clusters=3, random_state=42)
engagment_metrics['cluster'] = kmeans.fit_predict(engagment_metrics[['Sessions Frequency', 'Sessions Duration (ms)','Total Traffic (Bytes)']])
engagment_metrics.head()

Sessions Frequency  Sessions Duration (ms)  \
MSISDN/Number                                               
3.360100e+10                    1                116720.0   
3.360100e+10                    1                181230.0   
3.360100e+10                    1                134969.0   
3.360101e+10                    1                 49878.0   
3.360101e+10                    2                 37104.0   

               Total Traffic (Bytes)  cluster  
MSISDN/Number                                  
3.360100e+10            8.786906e+08        2  
3.360100e+10            1.568596e+08        2  
3.360100e+10            5.959665e+08        2  
3.360101e+10            4.223207e+08        2  
3.360101e+10            1.457411e+09        0

In [21]:
#grouping
engagment_metrics.groupby(['cluster']).describe().transpose()

cluster                                  0             1             2
Sessions Frequency     count  1.603000e+04  1.000000e+00  9.082600e+04
                       mean   2.754460e+00  1.065000e+03  1.153645e+00
                       std    1.140445e+00           NaN  3.877689e-01
                       min    2.000000e+00  1.065000e+03  1.000000e+00
                       25%    2.000000e+00  1.065000e+03  1.000000e+00
                       50%    2.000000e+00  1.065000e+03  1.000000e+00
                       75%    3.000000e+00  1.065000e+03  1.000000e+00
                       max    1.800000e+01  1.065000e+03  4.000000e+00
Sessions Duration (ms) count  1.603000e+04  1.000000e+00  9.082600e+04
                       mean   3.110938e+05  7.244640e+07  1.170590e+05
                       std    3.899402e+05           NaN  9.152224e+04
                       min    1.823500e+04  7.244640e+07  7.142000e+03
                       25%    1.624950e+05  7.244640e+07  6.069275e+04
                       50%    2.472350e+05  7.244640e+07  9.187800e+04
                       75%    3.540700e+05  7.244640e+07  1.550618e+05
                       max    1.855375e+07  7.244640e+07  3.174824e+06
Total Traffic (Bytes)  count  1.603000e+04  1.000000e+00  9.082600e+04
                       mean   1.580980e+09  5.312489e+11  5.338817e+08
                       std    5.510252e+08           NaN  2.561819e+08
                       min    1.055942e+09  5.312489e+11  3.324901e+07
                       25%    1.217108e+09  5.312489e+11  3.165844e+08
                       50%    1.421893e+09  5.312489e+11  5.424587e+08
                       75%    1.730165e+09  5.312489e+11  7.515387e+08
                       max    8.846226e+09  5.312489e+11  1.055771e+09

### Compute the minimum, maximum, average & total non-normalized metrics for each cluster. Interpret your results visually with accompanying text explaining your findings.


In [22]:
import numpy as np

# Compute the minimum, maximum, average & total non-normalized metrics for each cluster
cluster_metrics = engagment_metrics.groupby('cluster').agg({
    'Sessions Frequency': ['min', 'max', 'mean', 'sum'],
    'Sessions Duration (ms)': ['min', 'max', 'mean', 'sum'],
    'Total Traffic (Bytes)': ['min', 'max', 'mean', 'sum']
}).fillna(0)

# Rename multi-level columns for clarity
cluster_metrics.columns = ['_'.join(col).strip() for col in cluster_metrics.columns.values]

# Reset index to make 'Cluster' a column
cluster_metrics.reset_index(inplace=True)

# Display the metrics for each cluster
cluster_metrics.head()

cluster  Sessions Frequency_min  Sessions Frequency_max  \
0        0                       2                      18   
1        1                    1065                    1065   
2        2                       1                       4   

   Sessions Frequency_mean  Sessions Frequency_sum  \
0                 2.754460                   44154   
1              1065.000000                    1065   
2                 1.153645                  104781   

   Sessions Duration (ms)_min  Sessions Duration (ms)_max  \
0                     18235.0                  18553754.0   
1                  72446396.0                  72446396.0   
2                      7142.0                   3174824.0   

   Sessions Duration (ms)_mean  Sessions Duration (ms)_sum  \
0                 3.110938e+05                4.986834e+09   
1                 7.244640e+07                7.244640e+07   
2                 1.170590e+05                1.063200e+10   

   Total Traffic (Bytes)_min  Total Traffic (Bytes)_max  \
0               1.055942e+09               8.846226e+09   
1               5.312489e+11               5.312489e+11   
2               3.324901e+07               1.055771e+09   

   Total Traffic (Bytes)_mean  Total Traffic (Bytes)_sum  
0                1.580980e+09               2.534311e+13  
1                5.312489e+11               5.312489e+11  
2                5.338817e+08               4.849034e+13

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
import pandas as pd

fit_transform(X)
# Assuming your DataFrame is named 'agg_metrics'

# Extract relevant columns for clustering
X = agg_metrics[['Sessions Frequency', 'Total Duration (s)', 'Total Traffic (Bytes)']]

# Standardize (normalize) the data
scaler = StandardScaler()
X_scaled = scaler
# Run k-means clustering (k=3)
kmeans = KMeans(n_clusters=3, random_state=42)
agg_metrics['Engagement Group'] = kmeans.fit_predict(X_scaled)

# # Display the engagement group for each customer
# # print(agg_metrics[['MSISDN/Number', 'Engagement Group']])
print(agg_metrics[['MSISDN/Number', 'Engagement Group']].head())


# # Optional: Display cluster centers and other k-means information
print("\nCluster Centers:")
print(pd.DataFrame(scaler.inverse_transform(kmeans.cluster_centers_), columns=X.columns))

# # Optional: Visualize the clusters (if you have a 3D plot library available)
# # import matplotlib.pyplot as plt
# # from mpl_toolkits.mplot3d import Axes3D
# # fig = plt.figure(figsize=(10, 7))
# # ax = fig.add_subplot(111, projection='3d')
# # ax.scatter(X_scaled[:, 0], X_scaled[:, 1], X_scaled[:, 2], c=agg_metrics['Engagement Group'], cmap='viridis')
# # plt.show()


In [ ]:
AttributeError: 'NoneType' object has no attribute 'split'

In [ ]:
import pandas as pd

# Assuming your DataFrame is named 'telecom_data'

# Calculate sessions frequency per MSISDN
sessions_frequency = telecom_data.groupby('MSISDN/Number')['Bearer Id'].count().reset_index()
sessions_frequency.rename(columns={'Bearer Id': 'Sessions Frequency'}, inplace=True)

# Calculate session duration in seconds per MSISDN
session_duration = telecom_data.groupby('MSISDN/Number')['Dur. (ms)'].sum() / 1000
session_duration = session_duration.reset_index()
session_duration.rename(columns={'Dur. (ms)': 'Total Duration (s)'}, inplace=True)

# Calculate session total traffic (download and upload) per MSISDN
session_total_traffic = telecom_data.groupby('MSISDN/Number')['Total UL (Bytes)', 'Total DL (Bytes)'].sum()
session_total_traffic['Total Traffic (Bytes)'] = session_total_traffic['Total UL (Bytes)'] + session_total_traffic['Total DL (Bytes)']
session_total_traffic.reset_index(inplace=True)

# Merge the aggregated metrics into a single DataFrame
agg_metrics = pd.merge(sessions_frequency, session_duration, on='MSISDN/Number', how='outer')
agg_metrics = pd.merge(agg_metrics, session_total_traffic[['MSISDN/Number', 'Total Traffic (Bytes)']], on='MSISDN/Number', how='outer')

# Display the aggregated metrics per MSISDN
print(agg_metrics.head())

# Report the top 10 customers for each metric
top_10_sessions_frequency = agg_metrics.sort_values(by='Sessions Frequency', ascending=False).head(10)
top_10_duration = agg_metrics.sort_values(by='Total Duration (s)', ascending=False).head(10)
top_10_total_traffic = agg_metrics.sort_values(by='Total Traffic (Bytes)', ascending=False).head(10)

# Display the top 10 customers for each metric
print("Top 10 Customers by Sessions Frequency:")
print(top_10_sessions_frequency[['MSISDN/Number', 'Sessions Frequency']])

print("\nTop 10 Customers by Total Duration:")
print(top_10_duration[['MSISDN/Number', 'Total Duration (s)']])

print("\nTop 10 Customers by Total Traffic:")
print(top_10_total_traffic[['MSISDN/Number', 'Total Traffic (Bytes)']])
